src/font.rs

```diff
@@ -114,6 +114,273 @@ impl<'a> Loca<'a> {
+    }
+
+    fn get_length(&'a self) -> u16 {
+        let &Encoding(data) = self;
+        get_u16(data, 2).unwrap()
+    }
+
+    fn get_language(&'a self) -> u16 {
+        let &Encoding(data) = self;
+        get_u16(data, 4).unwrap()
+    }
+
+    fn get_seg_count_x_2(&'a self) -> u16 {
+        let &Encoding(data) = self;
+        get_u16(data, 6).unwrap()
+    }
+
+    fn get_seg_count(&'a self) -> u16 {
+        self.get_seg_count_x_2() / 2
+    }
+
+    fn get_search_range(&'a self) -> u16 {
+        let &Encoding(data) = self;
+        get_u16(data, 8).unwrap()
+    }
+
+    fn get_entry_selector(&'a self) -> u16 {
+        let &Encoding(data) = self;
+        get_u16(data, 10).unwrap()
+    }
+
+    fn get_range_shift(&'a self) -> u16 {
+        let &Encoding(data) = self;
+        get_u16(data, 12).unwrap()
+    }
+
+    fn get_u16_vec(&'a self, start_position: u16, count: u16) -> Vec<u16> {
+        let &Encoding(data) = self;
+        let mut result = vec![];
+        let mut vec_position = start_position;
+        let limit = vec_position + 2 * count;
+        while vec_position < limit {
+            result.push(get_u16(data, vec_position as usize).unwrap());
+            vec_position += 2;
+        }
+        result
+    }
+
+    fn get_i16_vec(&'a self, start_position: u16, count: u16) -> Vec<i16> {
+        let &Encoding(data) = self;
+        let mut result = vec![];
+        let mut vec_position = start_position;
+        let limit = vec_position + 2 * count;
+        while vec_position < limit {
+            result.push(get_i16(data, vec_position as usize).unwrap());
+            vec_position += 2;
+        }
+        result
+    }
+
+    fn get_end_counts_position() -> u16 {
+        14
+    }
+
+    fn get_end_counts(&'a self) -> Vec<u16> {
+        let seg_count = self.get_seg_count();
+        self.get_u16_vec(Self::get_end_counts_position(), seg_count)
+    }
+
+    fn get_start_counts_position(seg_count: u16) -> u16 {
+        Self::get_end_counts_position() + 2 + 2 * seg_count
+    }
+
+    fn get_start_counts(&'a self) -> Vec<u16> {
+        let seg_count = self.get_seg_count();
+        self.get_u16_vec(Self::get_start_counts_position(seg_count), seg_count)
+    }
+
+    fn get_id_deltas_position(seg_count: u16) -> u16 {
+        Self::get_start_counts_position(seg_count) + 2 * seg_count
+    }
+
+    fn get_id_deltas(&'a self) -> Vec<i16> {
+        let seg_count = self.get_seg_count();
+        self.get_i16_vec(Self::get_id_deltas_position(seg_count), seg_count)
+    }
+
+    fn get_id_range_offset_position(seg_count: u16) -> u16 {
+        Self::get_id_deltas_position(seg_count) + 2 * seg_count
+    }
+
+    fn get_id_range_offsets(&'a self) -> Vec<u16> {
+        let seg_count = self.get_seg_count();
+        self.get_u16_vec(Self::get_id_range_offset_position(seg_count), seg_count)
+    }
+
+    pub fn lookup_glyph_id(&'a self, code_point: u16) -> Option<u16> {
+        if self.get_format() != 4 {
```

Comment:
This doesn't feel like the right place for this. Ideally I'd like to see both an Encoding and an EncodingFormat4 type, with the creation of the latter gated on the format check. Thus, within an EncodingFormat4 impl, the invariant of the format can be guaranteed to hold. This will generalize much more easily to handling other formats.

Reply:
Agreed.

What relationship to you imagine existing between Encoding and EncodingFormat4?